examples/get_matching/server.c

```diff
@@ -0,0 +1,191 @@
+	long num_messages = 0;
+	char* p;
+
+	if (argc < 2) {
+		wait_for_ever = true;
+	} else {
+		num_messages = strtol(argv[1], &p, 10);
+		if (*p != '\0') {
+			fprintf(stderr,
+				"An invalid character was found before the end of the string");
+			return 1;
+		}
+	}
+
+	printf(" \n--- Serveur interface --- \n \n");
+
+	ret = PtlInit();
+	if (ret != PTL_OK) {
+		fprintf(stderr, "PtlInit failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+		return 1;
+	}
+
+	ret = PtlNIInit(PTL_IFACE_DEFAULT, PTL_NI_NO_MATCHING | PTL_NI_PHYSICAL, SERVER_PID, NULL,
+			NULL, &nih);
+	if (ret != PTL_OK) {
+		fprintf(stderr, "PtlNIInit failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+		res = 1;
+		goto fini;
+	}
+
+	ret = PtlEQAlloc(nih, 10, &eqh);
+	if (ret != PTL_OK) {
+		fprintf(stderr, "PtlEQAlloc failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+		res = 1;
+		goto ni_fini;
+	}
+
+	ret = PtlPTAlloc(nih, 0, eqh, 0, &pti);
+	if (ret != PTL_OK) {
+		fprintf(stderr, "PtlPTAlloc failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+		res = 1;
+		goto eq_free;
+	}
+
+	/* get server pid and server nid so the client know the server id */
+	ret = PtlGetId(nih, &id);
+	if (ret != PTL_OK) {
+		fprintf(stderr, "PtlGetId failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+		res = 1;
+		goto pt_free;
+	}
+	printf("Server Pid : %d \n", id.phys.pid);
+	printf("Server Nid : %d \n\n", id.phys.nid);
+
+	/* describe the memory used to send messages */
+	me = (ptl_me_t){ .start = data,
+			.length = sizeof(data),
+			.ct_handle = PTL_CT_NONE,
+			.uid = PTL_UID_ANY,
+			.options = PTL_LE_OP_GET | PTL_ME_MANAGE_LOCAL,
+			.match_id = id_server,
+			.match_bits = 0,
+			.ignore_bits = 0,
+			.min_free = MESSAGELENGTH-1 };
+
+	ret = PtlMEAppend(nih, pti, &me, 0, NULL, &meh);
+	if (ret != PTL_OK) {
+		fprintf(stderr, "PtlLEAppend failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+		res = 1;
+		goto pt_free;
+	}
+
+	ret = PtlEQWait(eqh, &ev);
+	if (ret != PTL_OK) {
+		fprintf(stderr, "PtlEQWait failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+		res = 1;
+		goto me_unlink1;
+	}
+	PtlEvToStr(0, &ev, msg);
+	printf("Event : %s \n", msg);
+
+	while(num_messages > 0 || wait_for_ever) {
+		ret = PtlEQWait(eqh, &ev);
+		if (ret != PTL_OK) {
+			fprintf(stderr, "PtlEQWait failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+			res = 1;
+			goto me_unlink1;
+		}
+
+		PtlEvToStr(0, &ev, msg);
+		printf("Event : %s \n", msg);
+
+		if (ev.type == PTL_EVENT_AUTO_UNLINK) {
+			n++;
+			char m[11];
+			sprintf(m, "message%d !", n);
+			memcpy(&data[0], m, 11);
+			ret = PtlMEAppend(nih, pti, &me, PTL_PRIORITY_LIST, NULL, &meh);
+			if (ret != PTL_OK) {
+				fprintf(stderr, "PtlLEAppend failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+				res = 1;
+				goto me_unlink1;
+			}
+
+			ret = PtlEQWait(eqh, &ev);
+			if (ret != PTL_OK) {
+				fprintf(stderr, "PtlEQWait failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+				res = 1;
+				goto me_unlink2;
+			}
+			PtlEvToStr(0, &ev, msg);
+			printf("Event : %s \n", msg);

```

Comment:
You could introduce an helper function, avoiding the repeat of the Append + Wait twice

Reply:
The two MEAppend do not have the same goto in case of error. I prefer keeping thinks like that.